python/tvm/relay/quantize/_annotate.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import absolute_import
+
+from . import _quantize
+from .quantize import QAnnotateKind, current_qconfig
+from .quantize import _conv_counter, _set_conv_counter
+from .. import expr as _expr
+from .. import op as _op
+from ..base import register_relay_node
+from ..._ffi.function import register_func
+
+
+@register_relay_node
+class QAnnotateExpr(_expr.TempExpr):
+    """A special kind of Expr for Annotating.
+
+    Parameters
+    ---------
+    expr: Expr
+        the original relay ir expr.
+
+    kind: QAnnotateKind
+        the kind of annotation field.
+    """
+    def __init__(self, expr, kind):
+        self.__init_handle_by_constructor__(
+            _quantize.make_annotate_expr, expr, kind)
+
+
+def _forward_op(ref_call, args):
+    """forward the operator of ref_call with provided arguments"""
+    return _expr.Call(
+        ref_call.op, args, ref_call.attrs, ref_call.type_args)
+
+
+def _get_expr_kind(anno):
+    """Get the expression and QAnnotateKind from QAnnotateExpr or Expr"""
+    if isinstance(anno, QAnnotateExpr):
+        return anno.expr, anno.kind
+    else:
+        return anno, None
+
+
+def register_annotate_function(op_name, frewrite=None, level=10):
+    """register a rewrite function for operator, used by annotation.
+
+    Parameters
+    ---------
+    op_name: str
+        The name of operation
+
+    frewrite : function, optional
+        The function to be registered.
+
+    level : int, optional
+        The priority level
+    """
+    return _op.register(op_name, "FQAnnotateRewrite", frewrite, level)
+
+
+@register_func("relay.quantize.attach_simulated_quantize")
+def attach_simulated_quantize(data, kind, sign=True, rounding="round"):
+    """Attach a simulated quantize operation after input data expr.
+
+    Parameters
+    ---------
+    data: Expr
+        the original data expr.
+
+    kind: QAnnotateKind
+        the kind of annotation field.
+    """
+    dom_scale = _expr.var("dom_scale")
+    clip_min = _expr.var("clip_min")
+    clip_max = _expr.var("clip_max")
+    return _quantize.simulated_quantize(
+        data, dom_scale, clip_min, clip_max, kind, sign, rounding)
+
+
+@register_annotate_function("nn.conv2d")
+def conv2d_rewrite(ref_call, new_args, ctx):
+    cnt = _conv_counter()
+    if cnt < current_qconfig().skip_k_conv:
+        _set_conv_counter(cnt + 1)
+        return None
+    _set_conv_counter(cnt + 1)
+
+    lhs_expr, lhs_kind = _get_expr_kind(new_args[0])
+    rhs_expr, rhs_kind = _get_expr_kind(new_args[1])
+
+    if lhs_kind is None or lhs_kind != QAnnotateKind.INPUT:
+        lhs_expr = attach_simulated_quantize(lhs_expr, QAnnotateKind.INPUT)
```

Comment:
can / should we avoid duplicated quantization in parallel branches? e.g 
```
   data (fp32) 
    /       \
conv1    conv2
```

Reply:
we should and we can, possibly via memoization. In theory forward rewrite already memoize, if there is any problem, please provide a minimum test case and let us double check

Reply:
This is a test case, to reproduce, you need to set opt level  `"CombineParallelConv2D": 4` to disable this pass.
```
import tvm
import tvm.relay as relay
import tvm.relay.testing

def get_workload():
    data = relay.var("data", shape=(1, 3, 224, 224), dtype='float32')
    conv1 = relay.testing.layers.conv2d(data=data, channels=16, kernel_size=(1, 1), name='conv1')
    shortcut = relay.testing.layers.conv2d(data=data, channels=16, kernel_size=(1, 1), name='sc')
    net = relay.add(conv1, shortcut)
    f = relay.Function(relay.ir_pass.free_vars(net), net)
    return relay.testing.init.create_workload(f)
    
sym, params = get_workload()
with tvm.relay.quantize.qconfig(skip_k_conv=0):
    sym = relay.quantize.quantize(sym, params)
print(sym.astext(show_meta_data=False))
tvm.relay.build(sym, 'llvm', params=params)
```
Result:
```
fn (%data: Tensor[(1, 3, 224, 224), float32])
    -> Tensor[(1, 16, 224, 224), float32] {
  %0 = multiply(%data, 16f) # ty=Tensor[(1, 3, 224, 224), float32]
  %1 = round(%0) # ty=Tensor[(1, 3, 224, 224), float32]
  %2 = clip(%1, a_min=-127, a_max=127) # ty=Tensor[(1, 3, 224, 224), float32]
  %3 = cast(%2, dtype="int8") # ty=Tensor[(1, 3, 224, 224), int8]
  %4 = meta.relay.Constant(id=0) # ty=Tensor[(16, 3, 1, 1), int8]
  %5 = nn.conv2d(%3, %4, channels=16, kernel_size=[1, 1], out_dtype="int32")
  %6 = multiply(%data, 16f) # ty=Tensor[(1, 3, 224, 224), float32]
  %7 = round(%6) # ty=Tensor[(1, 3, 224, 224), float32]
  %8 = clip(%7, a_min=-127, a_max=127) # ty=Tensor[(1, 3, 224, 224), float32]
  %9 = cast(%8, dtype="int8") # ty=Tensor[(1, 3, 224, 224), int8]
  %10 = meta.relay.Constant(id=1) # ty=Tensor[(16, 3, 1, 1), int8]
  %11 = nn.conv2d(%9, %10, channels=16, kernel_size=[1, 1], out_dtype="int32")
  %12 = add(%5, %11)
  %13 = add(%12, 64)
  %14 = right_shift(%13, 7)
  %15 = clip(%14, a_min=-127, a_max=127)
  %16 = cast(%15, dtype="int8")
  %17 = cast(%16, dtype="float32")
  %18 = multiply(%17, 0.0625f)
  %18
}
```

Reply:
@ZihengJiang @merrymercy @vinx13 can you look into this? let us open this testcase as an issue to be fixed

Reply:
This is because quantization of data happens during rewrite of conv2d, so this won't be memorized. We need some message passing to quantize data during forward rewrite of data.

Reply:
In alter_op_layout, I met the same problem as vinx13 mentioned so I use an internal memorizer.

Reply:
I see, this seems to be a common enough pattern that worth documenting

Reply:
Couldn't we also avoid this by using ANF ?

Reply:
It does not have things to do with ANF. The problem is that if two conv refers to the same input and they want to run the same transformation f on that input, there will be two such f.

One solution is to build a generic common subexpression combination(elimination) path to create a concise dag

Reply:
As we have seen in @vinx13 's test case, there're three `multiply` operations. The multipliers are typically `16` or `1/16` (which is 0.0625). In order to eliminate floating-point multiplication, can we convert them into shift operation upon integers?

Reply:
As far as i understand, this PR already do that

Reply:
@ZihengJiang What's the status of this issue of parallel branches? Will it be future work?